Error `polygon.consensus.syncer: failed to notify` after adding + removing peers/validators

## Error `polygon.consensus.syncer: failed to notify` after adding + removing peers/validators

### Description

I've seen many occurrences of this new error after adding and removing peers and validators to the original validator set:
```
2021-09-10T16:32:00.056Z [ERROR] polygon.consensus.syncer: failed to notify: err="rpc error: code = Unavailable desc = connection error: desc = "transport: failed to write client preface: stream reset""
```

It seems as if the active nodes may somehow still maintain a reference to old (removed) peers and/or validators even so they are no longer visible when running the SDK commands:

```
./main peers list
```
and
```
./main ibft snapshot
```

I've made sure to run `./main ibft propose --vote drop` on greater than 50% of active validators (removing it from the list that `ibft snapshot` returns, as expected) but it seems I only see an even greater frequency of the same error -- for reference below, validators were removed at 12:08 and 12:14:
![Screen Shot 2021-09-10 at 12 40 42 PM](https://user-images.githubusercontent.com/3344880/132888100-68afb622-8ea7-4402-bf54-e9b58ef72631.png)


### Your environment

* OS and version `OSX Big Sur 11.5.2 (20G95)`
* version of the Polygon SDK `0b2cc26`
* branch that causes this issue `develop`

### Steps to reproduce

1. Run `ibft server` for 4 bootnodes
2. Add peer `new_node1` using `peers add`
3. Vote in new validator `new_node1` using `ibft propose --vote auth`
4. Vote in new validator `new_node2` using `ibft propose --vote auth`
5. Stop validators `new_node1`, `new_node2`
6. Drop both new validators `new_node1`, `new_node2` using `ibft propose --vote drop`
7. Verify that `peers list` shows 3 peers for each node (as expected for 4 bootnodes)
8. Verify that `ibft snapshot` shows 4 validator addresses total (as expected for 4 bootnodes)
9. See new errors in logs

### Expected behaviour

I shouldn't any of these errors.

### Actual behaviour

I see many of these errors in the logs on all 4 bootnodes.

I initially suspected that the validators were stopped prematurely without being dropped by the active validators but when attempting to remove them from the validator list, it seems that I only see more of the same error in the logs.

--- a/e2e/broadcast_test.go
+++ b/e2e/broadcast_test.go
@@ -105,7 +105,7 @@ func TestBroadcast(t *testing.T) {
 						assert.NoError(t, err)
 						assert.Equal(t, uint64(1), res.Length)
 					} else {
-						assert.ErrorIs(t, err, framework.ErrTimeout)
+						assert.ErrorIs(t, err, tests.ErrTimeout)
 					}
 				})
 			}

--- a/e2e/framework/helper.go
+++ b/e2e/framework/helper.go
@@ -13,6 +13,7 @@ import (
 	"time"
 
 	"github.com/0xPolygon/polygon-sdk/crypto"
+	"github.com/0xPolygon/polygon-sdk/helper/tests"
 	"github.com/0xPolygon/polygon-sdk/server/proto"
 	txpoolProto "github.com/0xPolygon/polygon-sdk/txpool/proto"
 	"github.com/0xPolygon/polygon-sdk/types"
@@ -103,7 +104,7 @@ func MultiJoin(t *testing.T, srvs ...*TestServer) {
 // otherwise returns timeout
 func WaitUntilPeerConnects(ctx context.Context, srv *TestServer, requiredNum int) (*proto.PeersListResponse, error) {
 	clt := srv.Operator()
-	res, err := RetryUntilTimeout(ctx, func() (interface{}, bool) {
+	res, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
 		subCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 		defer cancel()
 		res, _ := clt.PeersList(subCtx, &empty.Empty{})
@@ -123,7 +124,7 @@ func WaitUntilPeerConnects(ctx context.Context, srv *TestServer, requiredNum int
 // otherwise returns timeout
 func WaitUntilTxPoolFilled(ctx context.Context, srv *TestServer, requiredNum uint64) (*txpoolProto.TxnPoolStatusResp, error) {
 	clt := srv.TxnPoolOperator()
-	res, err := RetryUntilTimeout(ctx, func() (interface{}, bool) {
+	res, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
 		subCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 		defer cancel()
 		res, _ := clt.Status(subCtx, &empty.Empty{})
@@ -137,33 +138,6 @@ func WaitUntilTxPoolFilled(ctx context.Context, srv *TestServer, requiredNum uin
 		return nil, err
 	}
 	return res.(*txpoolProto.TxnPoolStatusResp), nil
-}
-
-func RetryUntilTimeout(ctx context.Context, f func() (interface{}, bool)) (interface{}, error) {
-	type result struct {
-		data interface{}
-		err  error
-	}
-	resCh := make(chan result, 1)
-	go func() {
-		defer close(resCh)
-		for {
-			select {
-			case <-ctx.Done():
-				resCh <- result{nil, ErrTimeout}
-				return
-			default:
-				res, retry := f()
-				if !retry {
-					resCh <- result{res, nil}
-					return
-				}
-			}
-			time.Sleep(time.Second)
-		}
-	}()
-	res := <-resCh
-	return res.data, res.err
 }
 
 // MethodSig returns the signature of a non-parametrized function

--- a/e2e/framework/testserver.go
+++ b/e2e/framework/testserver.go
@@ -21,6 +21,7 @@ import (
 	"github.com/0xPolygon/polygon-sdk/command/server"
 	"github.com/0xPolygon/polygon-sdk/consensus/ibft"
 	"github.com/0xPolygon/polygon-sdk/crypto"
+	"github.com/0xPolygon/polygon-sdk/helper/tests"
 	"github.com/0xPolygon/polygon-sdk/network"
 	"github.com/0xPolygon/polygon-sdk/server/proto"
 	txpoolProto "github.com/0xPolygon/polygon-sdk/txpool/proto"
@@ -37,10 +38,6 @@ type TestServerConfigCallback func(*TestServerConfig)
 const (
 	initialPort   = 12000
 	polygonSDKCmd = "polygon-sdk"
-)
-
-var (
-	ErrTimeout = errors.New("timeout")
 )
 
 type TestServer struct {
@@ -261,7 +258,7 @@ func (t *TestServer) Start(ctx context.Context) error {
 		return err
 	}
 
-	_, err := RetryUntilTimeout(ctx, func() (interface{}, bool) {
+	_, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
 		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 		defer cancel()
 
@@ -362,7 +359,7 @@ func (t *TestServer) WaitForReceipt(ctx context.Context, hash web3.Hash) (*web3.
 		err     error
 	}
 
-	res, err := RetryUntilTimeout(ctx, func() (interface{}, bool) {
+	res, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
 		receipt, err := client.Eth().GetTransactionReceipt(hash)
 		if err != nil && err.Error() != "not found" {
 			return result{receipt, err}, false
@@ -381,7 +378,7 @@ func (t *TestServer) WaitForReceipt(ctx context.Context, hash web3.Hash) (*web3.
 
 func (t *TestServer) WaitForReady(ctx context.Context) error {
 	client := t.JSONRPC()
-	_, err := RetryUntilTimeout(ctx, func() (interface{}, bool) {
+	_, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
 		num, err := client.Eth().BlockNumber()
 		if err != nil {
 			return nil, true

--- a/helper/tests/testing.go
+++ b/helper/tests/testing.go
@@ -1,12 +1,19 @@
 package tests
 
 import (
+	"context"
 	"crypto/ecdsa"
+	"errors"
 	"testing"
+	"time"
 
 	"github.com/0xPolygon/polygon-sdk/crypto"
 	"github.com/0xPolygon/polygon-sdk/types"
 	"github.com/stretchr/testify/assert"
+)
+
+var (
+	ErrTimeout = errors.New("timeout")
 )
 
 func GenerateKeyAndAddr(t *testing.T) (*ecdsa.PrivateKey, types.Address) {
@@ -15,4 +22,31 @@ func GenerateKeyAndAddr(t *testing.T) (*ecdsa.PrivateKey, types.Address) {
 	assert.NoError(t, err)
 	addr := crypto.PubKeyToAddress(&key.PublicKey)
 	return key, addr
+}
+
+func RetryUntilTimeout(ctx context.Context, f func() (interface{}, bool)) (interface{}, error) {
+	type result struct {
+		data interface{}
+		err  error
+	}
+	resCh := make(chan result, 1)
+	go func() {
+		defer close(resCh)
+		for {
+			select {
+			case <-ctx.Done():
+				resCh <- result{nil, ErrTimeout}
+				return
+			default:
+				res, retry := f()
+				if !retry {
+					resCh <- result{res, nil}
+					return
+				}
+			}
+			time.Sleep(time.Second)
+		}
+	}()
+	res := <-resCh
+	return res.data, res.err
 }

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -7,6 +7,7 @@ import (
 	"math"
 	"math/big"
 	"sync"
+	"sync/atomic"
 
 	"github.com/0xPolygon/polygon-sdk/blockchain"
 	"github.com/0xPolygon/polygon-sdk/network"
@@ -144,8 +145,8 @@ type Syncer struct {
 	logger     hclog.Logger
 	blockchain blockchainShim
 
-	peers     map[peer.ID]*syncPeer
-	peersLock sync.RWMutex
+	peers    sync.Map // Maps peer.ID -> syncPeer
+	peersNum int64    // Number of entries in the peers map
 
 	serviceV1 *serviceV1
 	stopCh    chan struct{}
@@ -160,7 +161,7 @@ type Syncer struct {
 func NewSyncer(logger hclog.Logger, server *network.Server, blockchain blockchainShim) *Syncer {
 	s := &Syncer{
 		logger:     logger.Named("syncer"),
-		peers:      map[peer.ID]*syncPeer{},
+		peersNum:   0,
 		stopCh:     make(chan struct{}),
 		blockchain: blockchain,
 		server:     server,
@@ -220,18 +221,15 @@ const syncerV1 = "/syncer/0.1"
 // enqueueBlock adds the specific block to the peerID queue
 func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 	s.logger.Debug("enqueue block", "peer", peerID, "number", b.Number(), "hash", b.Hash())
-	s.peersLock.RLock()
-	p, ok := s.peers[peerID]
-	s.peersLock.RUnlock()
+
+	foundPeer, ok := s.peers.Load(peerID)
 	if ok {
-		p.appendBlock(b)
+		foundPeer.(*syncPeer).appendBlock(b)
 	}
 }
 
 // Broadcast broadcasts a block to all peers
 func (s *Syncer) Broadcast(b *types.Block) {
-	s.peersLock.RLock()
-	defer s.peersLock.RUnlock()
 	// diff is number in ibft
 	diff := new(big.Int).SetUint64(b.Number())
 
@@ -246,11 +244,19 @@ func (s *Syncer) Broadcast(b *types.Block) {
 			Value: b.MarshalRLP(),
 		},
 	}
-	for _, p := range s.peers {
-		if _, err := p.client.Notify(context.Background(), req); err != nil {
+
+	s.peers.Range(func(peerID, foundPeer interface{}) bool {
+		if _, err := foundPeer.(*syncPeer).client.Notify(context.Background(), req); err != nil {
 			s.logger.Error("failed to notify", "err", err)
 		}
-	}
+
+		return true
+	})
+}
+
+// NumSyncPeers returns the number of peers in the sync map. [Thread safe]
+func (s *Syncer) NumSyncPeers() int64 {
+	return atomic.LoadInt64(&s.peersNum)
 }
 
 // Start starts the syncer protocol
@@ -281,9 +287,8 @@ func (s *Syncer) Start() {
 			}
 
 			if evnt.Type == network.PeerEventDisconnected {
-				s.peersLock.Lock()
-				delete(s.peers, evnt.PeerID)
-				s.peersLock.Unlock()
+				s.peers.Delete(evnt.PeerID)
+				atomic.AddInt64(&s.peersNum, -1)
 				continue
 			}
 			if evnt.Type != network.PeerEventConnected {
@@ -306,28 +311,31 @@ func (s *Syncer) Start() {
 func (s *Syncer) BestPeer() *syncPeer {
 	var bestPeer *syncPeer
 	var bestTd *big.Int
-	s.peersLock.RLock()
-	defer s.peersLock.RUnlock()
-	for _, p := range s.peers {
-		status := p.status
+
+	s.peers.Range(func(peerID, foundPeer interface{}) bool {
+		status := foundPeer.(*syncPeer).status
 		if bestPeer == nil || status.Difficulty.Cmp(bestTd) > 0 {
-			bestPeer, bestTd = p, status.Difficulty
+			bestPeer, bestTd = foundPeer.(*syncPeer), status.Difficulty
 		}
-	}
+
+		return true
+	})
+
 	if bestPeer == nil {
 		return nil
 	}
+
 	curDiff := s.blockchain.CurrentTD()
+
 	if bestTd.Cmp(curDiff) <= 0 {
 		return nil
 	}
+
 	return bestPeer
 }
 
 // HandleUser is a helper method that is used to handle new user connections within the Syncer
 func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
-	s.peersLock.Lock()
-	defer s.peersLock.Unlock()
 	// watch for changes of the other node first
 	clt := proto.NewV1Client(conn)
 
@@ -339,12 +347,15 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 	if err != nil {
 		return err
 	}
-	s.peers[peerID] = &syncPeer{
+
+	s.peers.Store(peerID, &syncPeer{
 		peer:      peerID,
 		client:    clt,
 		status:    status,
 		enqueueCh: make(chan struct{}),
-	}
+	})
+	atomic.AddInt64(&s.peersNum, 1)
+
 	return nil
 }
 

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -144,7 +144,8 @@ type Syncer struct {
 	logger     hclog.Logger
 	blockchain blockchainShim
 
-	peers map[peer.ID]*syncPeer // TODO: Remove
+	peers     map[peer.ID]*syncPeer
+	peersLock sync.RWMutex
 
 	serviceV1 *serviceV1
 	stopCh    chan struct{}
@@ -219,8 +220,9 @@ const syncerV1 = "/syncer/0.1"
 // enqueueBlock adds the specific block to the peerID queue
 func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 	s.logger.Debug("enqueue block", "peer", peerID, "number", b.Number(), "hash", b.Hash())
-
+	s.peersLock.RLock()
 	p, ok := s.peers[peerID]
+	s.peersLock.RUnlock()
 	if ok {
 		p.appendBlock(b)
 	}
@@ -228,6 +230,8 @@ func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 
 // Broadcast broadcasts a block to all peers
 func (s *Syncer) Broadcast(b *types.Block) {
+	s.peersLock.RLock()
+	defer s.peersLock.RUnlock()
 	// diff is number in ibft
 	diff := new(big.Int).SetUint64(b.Number())
 
@@ -276,6 +280,12 @@ func (s *Syncer) Start() {
 				return
 			}
 
+			if evnt.Type == network.PeerEventDisconnected {
+				s.peersLock.Lock()
+				delete(s.peers, evnt.PeerID)
+				s.peersLock.Unlock()
+				continue
+			}
 			if evnt.Type != network.PeerEventConnected {
 				continue
 			}
@@ -296,7 +306,8 @@ func (s *Syncer) Start() {
 func (s *Syncer) BestPeer() *syncPeer {
 	var bestPeer *syncPeer
 	var bestTd *big.Int
-
+	s.peersLock.RLock()
+	defer s.peersLock.RUnlock()
 	for _, p := range s.peers {
 		status := p.status
 		if bestPeer == nil || status.Difficulty.Cmp(bestTd) > 0 {
@@ -315,6 +326,8 @@ func (s *Syncer) BestPeer() *syncPeer {
 
 // HandleUser is a helper method that is used to handle new user connections within the Syncer
 func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
+	s.peersLock.Lock()
+	defer s.peersLock.Unlock()
 	// watch for changes of the other node first
 	clt := proto.NewV1Client(conn)
 

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -7,7 +7,6 @@ import (
 	"math"
 	"math/big"
 	"sync"
-	"sync/atomic"
 
 	"github.com/0xPolygon/polygon-sdk/blockchain"
 	"github.com/0xPolygon/polygon-sdk/network"
@@ -170,8 +169,7 @@ type Syncer struct {
 	logger     hclog.Logger
 	blockchain blockchainShim
 
-	peers    sync.Map // Maps peer.ID -> syncPeer
-	peersNum int64    // Number of entries in the peers map
+	peers sync.Map // Maps peer.ID -> syncPeer
 
 	serviceV1 *serviceV1
 	stopCh    chan struct{}
@@ -186,7 +184,6 @@ type Syncer struct {
 func NewSyncer(logger hclog.Logger, server *network.Server, blockchain blockchainShim) *Syncer {
 	s := &Syncer{
 		logger:     logger.Named("syncer"),
-		peersNum:   0,
 		stopCh:     make(chan struct{}),
 		blockchain: blockchain,
 		server:     server,
@@ -287,11 +284,6 @@ func (s *Syncer) Broadcast(b *types.Block) {
 	})
 }
 
-// NumSyncPeers returns the number of peers in the sync map. [Thread safe]
-func (s *Syncer) NumSyncPeers() int64 {
-	return atomic.LoadInt64(&s.peersNum)
-}
-
 // Start starts the syncer protocol
 func (s *Syncer) Start() {
 	s.serviceV1 = &serviceV1{syncer: s, logger: hclog.NewNullLogger(), store: s.blockchain}
@@ -387,7 +379,6 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 		status:    status,
 		enqueueCh: make(chan struct{}),
 	})
-	atomic.AddInt64(&s.peersNum, 1)
 
 	return nil
 }
@@ -395,7 +386,6 @@ func (s *Syncer) HandleUser(peerID peer.ID, conn *grpc.ClientConn) error {
 func (s *Syncer) DeleteUser(peerID peer.ID) error {
 	p, ok := s.peers.LoadAndDelete(peerID)
 	if ok {
-		atomic.AddInt64(&s.peersNum, -1)
 		if err := p.(*syncPeer).conn.Close(); err != nil {
 			return err
 		}

--- a/protocol/syncer.go
+++ b/protocol/syncer.go
@@ -244,17 +244,26 @@ const syncerV1 = "/syncer/0.1"
 func (s *Syncer) enqueueBlock(peerID peer.ID, b *types.Block) {
 	s.logger.Debug("enqueue block", "peer", peerID, "number", b.Number(), "hash", b.Hash())
 
-	foundPeer, ok := s.peers.Load(peerID)
+	peer, ok := s.peers.Load(peerID)
 	if ok {
-		foundPeer.(*syncPeer).appendBlock(b)
+		peer.(*syncPeer).appendBlock(b)
 	}
 }
 
 func (s *Syncer) updatePeerStatus(peerID peer.ID, status *Status) {
-	s.logger.Debug("update peer status", "peer", peerID, "latest block number", status.Number, "latest block hash", status.Hash, "difficulty", status.Difficulty)
+	s.logger.Debug(
+		"update peer status",
+		"peer",
+		peerID,
+		"latest block number",
+		status.Number,
+		"latest block hash",
+		status.Hash, "difficulty",
+		status.Difficulty,
+	)
 
-	if p, ok := s.peers.Load(peerID); ok {
-		p.(*syncPeer).updateStatus(status)
+	if peer, ok := s.peers.Load(peerID); ok {
+		peer.(*syncPeer).updateStatus(status)
 	}
 }
 
@@ -275,8 +284,8 @@ func (s *Syncer) Broadcast(b *types.Block) {
 		},
 	}
 
-	s.peers.Range(func(peerID, foundPeer interface{}) bool {
-		if _, err := foundPeer.(*syncPeer).client.Notify(context.Background(), req); err != nil {
+	s.peers.Range(func(peerID, peer interface{}) bool {
+		if _, err := peer.(*syncPeer).client.Notify(context.Background(), req); err != nil {
 			s.logger.Error("failed to notify", "err", err)
 		}
 
@@ -336,10 +345,10 @@ func (s *Syncer) BestPeer() *syncPeer {
 	var bestPeer *syncPeer
 	var bestTd *big.Int
 
-	s.peers.Range(func(peerID, foundPeer interface{}) bool {
-		status := foundPeer.(*syncPeer).status
+	s.peers.Range(func(peerID, peer interface{}) bool {
+		status := peer.(*syncPeer).status
 		if bestPeer == nil || status.Difficulty.Cmp(bestTd) > 0 {
-			bestPeer, bestTd = foundPeer.(*syncPeer), status.Difficulty
+			bestPeer, bestTd = peer.(*syncPeer), status.Difficulty
 		}
 
 		return true

--- a/protocol/syncer_test.go
+++ b/protocol/syncer_test.go
@@ -1,0 +1,140 @@
+package protocol
+
+import (
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/0xPolygon/polygon-sdk/blockchain"
+	"github.com/0xPolygon/polygon-sdk/network"
+	"github.com/0xPolygon/polygon-sdk/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/assert"
+)
+
+type mockBlockStore struct {
+	blocks       []*types.Block
+	subscription *blockchain.MockSubscription
+	td           *big.Int
+}
+
+func newMockBlockStore() *mockBlockStore {
+	bs := &mockBlockStore{
+		blocks:       make([]*types.Block, 0),
+		subscription: blockchain.NewMockSubscription(),
+		td:           big.NewInt(1),
+	}
+	return bs
+}
+
+func (m *mockBlockStore) Header() *types.Header {
+	return m.blocks[len(m.blocks)-1].Header
+}
+func (m *mockBlockStore) GetHeaderByNumber(n uint64) (*types.Header, bool) {
+	b, ok := m.GetBlockByNumber(n, false)
+	if !ok {
+		return nil, false
+	}
+	return b.Header, true
+}
+func (m *mockBlockStore) GetBlockByNumber(blockNumber uint64, full bool) (*types.Block, bool) {
+	for _, b := range m.blocks {
+		if b.Number() == blockNumber {
+			return b, true
+		}
+	}
+	return nil, false
+}
+func (m *mockBlockStore) SubscribeEvents() blockchain.Subscription {
+	return m.subscription
+}
+func (m *mockBlockStore) GetReceiptsByHash(types.Hash) ([]*types.Receipt, error) {
+	return nil, nil
+}
+
+func (m *mockBlockStore) GetHeaderByHash(hash types.Hash) (*types.Header, bool) {
+	for _, b := range m.blocks {
+		header := b.Header.ComputeHash()
+		if header.Hash == hash {
+			return header, true
+		}
+	}
+	return nil, true
+}
+func (m *mockBlockStore) GetBodyByHash(hash types.Hash) (*types.Body, bool) {
+	for _, b := range m.blocks {
+
+		if b.Hash() == hash {
+			return b.Body(), true
+		}
+	}
+	return nil, true
+}
+func (m *mockBlockStore) WriteBlocks(blocks []*types.Block) error {
+
+	for _, b := range blocks {
+		m.td.Add(m.td, big.NewInt(int64(b.Header.Difficulty)))
+		m.blocks = append(m.blocks, b)
+	}
+	return nil
+}
+
+func (m *mockBlockStore) CurrentTD() *big.Int {
+	return m.td
+}
+
+func (m *mockBlockStore) GetTD(hash types.Hash) (*big.Int, bool) {
+	return m.td, false
+}
+func createGenesisBlock() []*types.Block {
+	blocks := make([]*types.Block, 0)
+	genesis := &types.Header{Difficulty: 1, Number: 0}
+	genesis.ComputeHash()
+	b := &types.Block{
+		Header: genesis,
+	}
+	blocks = append(blocks, b)
+	return blocks
+}
+
+func createBlockStores(count int) (bStore []*mockBlockStore) {
+	bStore = make([]*mockBlockStore, count)
+	for i := 0; i < count; i++ {
+		bStore[i] = newMockBlockStore()
+	}
+	return
+}
+func TestSyncer_PeerDisconnected(t *testing.T) {
+	conf := func(c *network.Config) {
+		c.MaxPeers = 4
+		c.NoDiscover = true
+	}
+	blocks := createGenesisBlock()
+	// Create three servers
+	srv0 := network.CreateServer(t, conf)
+	srv1 := network.CreateServer(t, conf)
+	srv2 := network.CreateServer(t, conf)
+
+	bstores := createBlockStores(3)
+
+	bstores[0].WriteBlocks(blocks)
+	bstores[1].WriteBlocks(blocks)
+	bstores[2].WriteBlocks(blocks)
+
+	sync0 := NewSyncer(hclog.NewNullLogger(), srv0, bstores[0])
+	sync1 := NewSyncer(hclog.NewNullLogger(), srv1, bstores[1])
+	sync2 := NewSyncer(hclog.NewNullLogger(), srv2, bstores[2])
+
+	go sync0.Start()
+	go sync1.Start()
+	go sync2.Start()
+
+	network.MultiJoin(t, srv0, srv1, srv0, srv2, srv1, srv2)
+	time.Sleep(1 * time.Second)
+	//Disconnect peer peer2
+	srv1.Disconnect(srv2.AddrInfo().ID, "testing")
+	time.Sleep(1 * time.Second)
+
+	assert.NotContains(t, sync1.peers, srv2.AddrInfo().ID)
+
+}

--- a/protocol/syncer_test.go
+++ b/protocol/syncer_test.go
@@ -117,10 +117,12 @@ func TestSyncer_PeerDisconnected(t *testing.T) {
 
 	bstores := createBlockStores(3)
 
-	bstores[0].WriteBlocks(blocks)
-	bstores[1].WriteBlocks(blocks)
-	bstores[2].WriteBlocks(blocks)
-
+	err := bstores[0].WriteBlocks(blocks)
+	assert.NoError(t, err)
+	err = bstores[1].WriteBlocks(blocks)
+	assert.NoError(t, err)
+	err = bstores[2].WriteBlocks(blocks)
+	assert.NoError(t, err)
 	sync0 := NewSyncer(hclog.NewNullLogger(), srv0, bstores[0])
 	sync1 := NewSyncer(hclog.NewNullLogger(), srv1, bstores[1])
 	sync2 := NewSyncer(hclog.NewNullLogger(), srv2, bstores[2])
@@ -131,10 +133,10 @@ func TestSyncer_PeerDisconnected(t *testing.T) {
 
 	network.MultiJoin(t, srv0, srv1, srv0, srv2, srv1, srv2)
 	time.Sleep(1 * time.Second)
-	//Disconnect peer peer2
+	//Disconnect peer2
 	srv1.Disconnect(srv2.AddrInfo().ID, "testing")
 	time.Sleep(1 * time.Second)
-
+	//server1 syncer should not have disconnected peer
 	assert.NotContains(t, sync1.peers, srv2.AddrInfo().ID)
 
 }

--- a/protocol/syncer_test.go
+++ b/protocol/syncer_test.go
@@ -201,6 +201,8 @@ func TestSyncer_PeerDisconnected(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Unable to add sync peers, %v", err)
 	}
+	// Make sure the number of peers is correct
+	// -1 to exclude the current node
 	assert.Equal(t, int64(len(servers)-1), numPeers)
 
 	// Disconnect peer2
@@ -213,9 +215,14 @@ func TestSyncer_PeerDisconnected(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Unable to disconnect sync peers, %v", err)
 	}
+	// Make sure a single peer disconnected
+	// Additional -1 to exclude the current node
 	assert.Equal(t, int64(len(servers)-2), numPeers)
 
 	// server1 syncer should have disconnected from server2 peer
 	_, found := syncers[1].peers.Load(peerToDisconnect)
+
+	// Make sure that the disconnected peer is not in the
+	// reference node's sync peer map
 	assert.False(t, found)
 }

--- a/protocol/syncer_test.go
+++ b/protocol/syncer_test.go
@@ -130,10 +130,22 @@ func createSyncers(count int, servers []*network.Server, blockStores []*mockBloc
 	return syncers
 }
 
+// numSyncPeers returns the number of sync peers
+func numSyncPeers(syncer *Syncer) int64 {
+	num := 0
+	syncer.peers.Range(func(key, value interface{}) bool {
+		num++
+
+		return true
+	})
+
+	return int64(num)
+}
+
 // WaitUntilSyncPeersNumber waits until the number of sync peers reaches a certain number, otherwise it times out
 func WaitUntilSyncPeersNumber(ctx context.Context, syncer *Syncer, requiredNum int64) (int64, error) {
 	res, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
-		numPeers := syncer.NumSyncPeers()
+		numPeers := numSyncPeers(syncer)
 		if numPeers == requiredNum {
 			return numPeers, false
 		}

--- a/protocol/syncer_test.go
+++ b/protocol/syncer_test.go
@@ -1,11 +1,13 @@
 package protocol
 
 import (
+	"context"
 	"math/big"
 	"testing"
 	"time"
 
 	"github.com/0xPolygon/polygon-sdk/blockchain"
+	"github.com/0xPolygon/polygon-sdk/helper/tests"
 	"github.com/0xPolygon/polygon-sdk/network"
 	"github.com/0xPolygon/polygon-sdk/types"
 	"github.com/hashicorp/go-hclog"
@@ -104,39 +106,104 @@ func createBlockStores(count int) (bStore []*mockBlockStore) {
 	}
 	return
 }
+
+// createNetworkServers is a helper function for generating network servers
+func createNetworkServers(count int, t *testing.T, conf func(c *network.Config)) []*network.Server {
+	networkServers := make([]*network.Server, count)
+
+	for indx := 0; indx < count; indx++ {
+		networkServers[indx] = network.CreateServer(t, conf)
+	}
+
+	return networkServers
+}
+
+// createSyncers is a helper function for generating syncers. Servers and BlockStores should be at least the length
+// of count
+func createSyncers(count int, servers []*network.Server, blockStores []*mockBlockStore) []*Syncer {
+	syncers := make([]*Syncer, count)
+
+	for indx := 0; indx < count; indx++ {
+		syncers[indx] = NewSyncer(hclog.NewNullLogger(), servers[indx], blockStores[indx])
+	}
+
+	return syncers
+}
+
+// WaitUntilSyncPeersNumber waits until the number of sync peers reaches a certain number, otherwise it times out
+func WaitUntilSyncPeersNumber(ctx context.Context, syncer *Syncer, requiredNum int64) (int64, error) {
+	res, err := tests.RetryUntilTimeout(ctx, func() (interface{}, bool) {
+		numPeers := syncer.NumSyncPeers()
+		if numPeers == requiredNum {
+			return numPeers, false
+		}
+		return nil, true
+	})
+
+	if err != nil {
+		return 0, err
+	}
+	return res.(int64), nil
+}
+
 func TestSyncer_PeerDisconnected(t *testing.T) {
 	conf := func(c *network.Config) {
 		c.MaxPeers = 4
 		c.NoDiscover = true
 	}
 	blocks := createGenesisBlock()
+
 	// Create three servers
-	srv0 := network.CreateServer(t, conf)
-	srv1 := network.CreateServer(t, conf)
-	srv2 := network.CreateServer(t, conf)
+	servers := createNetworkServers(3, t, conf)
 
-	bstores := createBlockStores(3)
+	// Create the block stores
+	blockStores := createBlockStores(3)
 
-	err := bstores[0].WriteBlocks(blocks)
-	assert.NoError(t, err)
-	err = bstores[1].WriteBlocks(blocks)
-	assert.NoError(t, err)
-	err = bstores[2].WriteBlocks(blocks)
-	assert.NoError(t, err)
-	sync0 := NewSyncer(hclog.NewNullLogger(), srv0, bstores[0])
-	sync1 := NewSyncer(hclog.NewNullLogger(), srv1, bstores[1])
-	sync2 := NewSyncer(hclog.NewNullLogger(), srv2, bstores[2])
+	for _, blockStore := range blockStores {
+		assert.NoError(t, blockStore.WriteBlocks(blocks))
+	}
 
-	go sync0.Start()
-	go sync1.Start()
-	go sync2.Start()
+	// Create the syncers
+	syncers := createSyncers(3, servers, blockStores)
 
-	network.MultiJoin(t, srv0, srv1, srv0, srv2, srv1, srv2)
-	time.Sleep(1 * time.Second)
-	//Disconnect peer2
-	srv1.Disconnect(srv2.AddrInfo().ID, "testing")
-	time.Sleep(1 * time.Second)
-	//server1 syncer should not have disconnected peer
-	assert.NotContains(t, sync1.peers, srv2.AddrInfo().ID)
+	// Start the syncers
+	for _, syncer := range syncers {
+		go syncer.Start()
+	}
 
+	network.MultiJoin(
+		t,
+		servers[0],
+		servers[1],
+		servers[0],
+		servers[2],
+		servers[1],
+		servers[2],
+	)
+
+	// wait until gossip protocol builds the mesh network (https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
+	waitCtx, cancelWait := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancelWait()
+
+	numPeers, err := WaitUntilSyncPeersNumber(waitCtx, syncers[1], 2)
+	if err != nil {
+		t.Fatalf("Unable to add sync peers, %v", err)
+	}
+	assert.Equal(t, int64(len(servers)-1), numPeers)
+
+	// Disconnect peer2
+	peerToDisconnect := servers[2].AddrInfo().ID
+	servers[1].Disconnect(peerToDisconnect, "testing")
+
+	waitCtx, cancelWait = context.WithTimeout(context.Background(), time.Second*10)
+	defer cancelWait()
+	numPeers, err = WaitUntilSyncPeersNumber(waitCtx, syncers[1], 1)
+	if err != nil {
+		t.Fatalf("Unable to disconnect sync peers, %v", err)
+	}
+	assert.Equal(t, int64(len(servers)-2), numPeers)
+
+	// server1 syncer should have disconnected from server2 peer
+	_, found := syncers[1].peers.Load(peerToDisconnect)
+	assert.False(t, found)
 }

